### PR TITLE
fix: templates plugins

### PR DESCRIPTION
**Pull request by dkonieczek**

### CODE REVIEW

**Comment by dkonieczek** [comment]
Some things I didn't include:
- ability to only apply middleware to specific features from within the config (instead of within the middleware)
- Tests since SnapTemplates still can't be tested (LibraryStore & swiper imports)
- Dynamic import/chunking of middleware (shopify updateResultsUrl) can't be done since plugins are attached to the config. Also since it's used at afterStore, it may delay the initial render if chunked?

File: packages/snap-preact/src/Templates/Stores/LibraryStore.ts
Comment (korgon) [inline packages/snap-preact/src/Templates/Stores/LibraryStore.ts:1]: I think we need to explore the dynamic import of plugins further. Without that functionality we will be bloating all Snap Templates integrations initial `bundle.js` file.

It is possible to attach plugins _after_ controller instantiation via the `controller.plugin` method - I am a little concerned about race conditions though with regards to dynamically importing a plugin and ensuring that it attaches prior to when it is needed.

File: packages/snap-preact/src/Templates/Stores/LibraryStore.ts [hunk 1]
```diff
@@ -16,6 +23,23 @@ export type LibraryImports = {
...
+			backgroundFilters: PluginFunction;
+			updateResultsUrl: PluginFunction;
+		};
+		bigcommerce: {
+			backgroundFilters: PluginFunction;
+		};
+		magento2: {
+			backgroundFilters: PluginFunction;
+		};
+		other: {
```
Comment (korgon) [inline packages/snap-preact/src/Templates/Stores/LibraryStore.ts:37]: I'd suggest `other` -> `common` ... but that is neither here nor there.

File: packages/snap-preact/src/Templates/Stores/library/middleware/genericBackgroundFilters.ts
Comment (korgon) [inline packages/snap-preact/src/Templates/Stores/library/middleware/genericBackgroundFilters.ts:1]: Technically not 'middleware' but 'plugins' (for the directory name).

File: docs/TEMPLATES_CONFIG.md [hunk 2]
```diff
@@ -86,6 +86,123 @@ It is possible to switch language and currency at run-time using methods on the
...
+|----------------------|-------------|------|---------|
+| `platform[platform].backgroundFilters` | Background filter configurations | Object | - |
+| `platform[platform].backgroundFilters.tags` | Background tag filter configurations | Array | - |
+| `platform[platform].backgroundFilters.collection` | Background collection filter configurations | Array | - |
+| `platform[platform].backgroundFilters.other` | Background filter configurations for custom fields | Array | - |
+| `platform[platform].backgroundFilters['tags' \| 'collection' \| 'other'][]` | Background filter object | Object | - |
+| `platform[platform].backgroundFilters['tags' \| 'collection' \| 'other'][].enabled` | Enables filter | boolean | false |
+| `platform[platform].backgroundFilters['tags' \| 'collection' \| 'other'][].field` | Defines filter field name | string | - |
+| `platform[platform].backgroundFilters['tags' \| 'collection' \| 'other'][].value` | Defines filter value | string | - |
+| `platform[platform].backgroundFilters['tags' \| 'collection' \| 'other'][].background` | Defines if filter should be background | boolean | true |
```
Comment (korgon) [inline docs/TEMPLATES_CONFIG.md:114]: Not sure why 'tags' and 'collection' and 'other' are here... 😕

File: packages/snap-preact/src/Templates/Stores/LibraryStore.ts
Comment (korgon) [reply packages/snap-preact/src/Templates/Stores/LibraryStore.ts:1]: This still needs to be figured out.

**Comment by korgon** [comment]
Much refactoring was done - but still some work to do on documentation and testing.

**Pull request merged**